### PR TITLE
Fixes GDScript define nested dictionary and array as constants

### LINKED ISSUE #50285: GDScript 2.0: Can't define nested dictionary and array constants
### Godot version

v4.0.dev.calinou [bfc14a635]

### System information

Windows 10

### Issue description

Upon defining a nested dictionary constant the editor will mark it as an error. It does, however, work when defining the inner dictionary (as a constant) beforehand. Not sure whether this is by design but it did work in GDScript 1.0 (Godot 3.x) so I'm guessing this may be a bug.

### Steps to reproduce

Define a nested dictionary as a constant inside a GDScript file (see below)

### Minimal reproduction project

Create a new GDScript file inside the engine and paste the following:
```
const CONST_DICTIONARY = {
	"key1": "value1",
	"key2": "value2",
}

const CONST_NESTED_DICTIONARY = { # Error
	"key1": "value1",
	"key2": { "key": true }
}

const CONST_INNER_DICTIONARY = { "key": true }
const CONST_NESTED_DICTIONARY_WORKAROUND = {
	"key1": "value1",
	"key2": CONST_INNER_DICTIONARY
}
```

### PR DESCRIPTION
**Pull request by PastMoments**
Fixes #50285, note the regression was introduced in (#41983) (commit 3886a2f9f6eabfe4dc32cd8eb9abe827cae9963d)

This issue seems to have been caused by the logic for checking const of elements being moved from `reduce_array` to `const_fold_array` (and `reduce_dict` to `const_fold_dict`) by the regression. Thus `reduce_expression` was not properly assigning const to these types.

This PR just adds extra checks in `const_fold_array` and `const_fold_dict` to handle nested arrays/dicts.

### CODE REVIEW

**Comment by akien-mga** [comment]
This seems to make a number of GDScript tests fail, and eventually crash: https://github.com/godotengine/godot/runs/5714947096?check_suite_focus=true

**Comment by PastMoments** [comment]
I've force pushed a new fix, and also added tests for the issue.

File: modules/gdscript/gdscript_analyzer.cpp [hunk 1]
```diff
@@ -3159,6 +3159,13 @@ void GDScriptAnalyzer::reduce_subscript(GDScriptParser::SubscriptNode *p_subscri
 		reduce_identifier(static_cast<GDScriptParser::IdentifierNode *>(p_subscript->base), true);
 	} else {
 		reduce_expression(p_subscript->base);
+
+		if (p_subscript->base->type == GDScriptParser::Node::ARRAY) {
+			GDScriptParser::ArrayNode *array = static_cast<GDScriptParser::ArrayNode *>(p_subscript->base);
```
Comment (vnen) [inline modules/gdscript/gdscript_analyzer.cpp:3164]: Any reason for this to use an extra variable while the other cases don't? It should work the same but I find this inconsistency odd, and probably so will the future people reading this code.

**Comment by vnen** [comment]
Overall looks okay to me, despite my comment. This needs a rebase to remove the merge commit though (see https://docs.godotengine.org/en/stable/community/contributing/pr_workflow.html#the-interactive-rebase).

**Comment by PastMoments** [comment]
@vnen Resolved the comment and rebased.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks! And congrats for your first merged Godot contribution :tada:

**Pull request merged**